Attempt to implement comments parsing for deserialize

jollm opened the pull request:
This change attempts to handle the following case:

"If a ; character is encountered outside of a string, that character
and all subsequent characters to the next newline should be ignored."

as described in https://github.com/edn-format/edn#comments

jollm committed:
Attempt to implement comments parsing for deserialize

This change attempts to handle the following case:

"If a ; character is encountered outside of a string, that character
and all subsequent characters to the next newline should be ignored."

as described in https://github.com/edn-format/edn#comments

naomijub commented on src/deserialize/parse.rs:
```diff
@@ -397,6 +426,10 @@ fn read_map(chars: &mut std::iter::Enumerate<std::str::Chars>) -> Result<Edn, Er
             Some((_, '}')) => return Ok(Edn::Map(Map::new(res))),
+            Some((_, ';')) => {
+                read_comment(chars)?;
+                ()
```
You don't need `()` here (Rust will interpret a line ending in `;` as `()`.

naomijub replied:
I do think this expression is not necessary, as it is already being considered in `parse`

jollm replied:
thanks for the tip on that!, I'll fix all of those similar cases

the top level parse doesn't catch inside a map, vec, set, etc., so if i comment this line the following test fails (from line 852):
```test deserialize::parse::test::parse_comment_inside_map ... FAILED```

naomijub commented on src/deserialize/parse.rs:
```diff
@@ -13,6 +13,14 @@ pub(crate) fn parse(
         Some((_, '{')) => read_map(chars)?,
+        Some((_, ';')) => {
+            let cmt = read_comment(chars)?;
+            match chars.next() {
```
Considering read_comment returns `Edn::Empty` and the functions receives a mutable chars, I don't think this match is necessary

jollm replied:
This is for the case where the only thing existing is a comment (tests starting on line 815) (fails without match). Normally the return from read_comment is ignored in favor of whatever the next call to parse returns. Your comment made me realize that an attempt to deserialize an empty or whitespace only string will fail instead of return Edn::Empty, so it would be surprising for comment to behave differently. For now I will make both return Edn::Empty, but the other option is to have both fail..

naomijub commented on src/deserialize/parse.rs:
```diff
@@ -26,8 +34,13 @@ pub(crate) fn parse_edn(
         Some((_, '\\')) => Ok(read_char(chars)?),
+        Some((_, ';')) => {
+            read_comment(chars)?;
+            Ok(parse_edn(chars.next(), chars)?)
```
if you call `parse_edn` here you will be missing the full parse (`parse`), as `parse_edn` only parses "primitive" types. Also, I am not sure you actually need this line `Ok(parse_edn(chars.next(), chars)?)`, maybe only if comments come as the first line of code.

jollm replied:
only parse calls parse_edn, so if parse_edn calls itself, it will still return back to parse?

Yes that is for comments as the first line, e.g. test parse_str_top_level_comment on line 494

naomijub replied:
`parse`  calls `parse_edn`, https://github.com/naomijub/edn-rs/blob/master/src/deserialize/parse.rs#L16.
I think this is the reason why 
> the top level parse doesn't catch inside a map, vec, set, etc., so if i comment this line the following test fails (from line 852): `test deserialize::parse::test::parse_comment_inside_map ... FAILED`

is failing

jollm replied:
It calls back to top level now.

jollm replied:
I see what you mean, this shouldn't be needed, I have removed these from parse_edn

naomijub commented on src/deserialize/parse.rs:
```diff
@@ -13,6 +13,14 @@ pub(crate) fn parse(
+                None => cmt,
+            }
+        }
+        Some((_, s)) if s.is_whitespace() => parse(chars.next(), chars)?,
```
I don't get why we need this extra parse? the code in general already discards whitespaces and commas.

jollm replied:
independent of the case with comments, if i comment that line, e.g. the following test will fail:
```
    #[test]
    fn parse_simple_map() {
        let mut edn = " {:a \"2\" :b false :c nil }".chars().enumerate();

        assert_eq!(
            parse(edn.next(), &mut edn).unwrap(),
            Edn::Map(Map::new(
                map! {":a".to_string() => Edn::Str("2".to_string()),
                ":b".to_string() => Edn::Bool(false), ":c".to_string() => Edn::Nil}
            ))
        );
    }
```
because of not handling whitespace at the top level (unless I'm missing where this is done at another level, which is entirely possible)

naomijub commented on src/deserialize/parse.rs:
```diff
@@ -328,6 +349,10 @@ fn read_set(chars: &mut std::iter::Enumerate<std::str::Chars>) -> Result<Edn, Er
             Some((_, '}')) => return Ok(Edn::Set(Set::new(res))),
+            Some((_, ';')) => {
+                read_comment(chars)?;
+                ()
```
You don't need () here (Rust will interpret a line ending in ; as ().

naomijub commented on src/deserialize/parse.rs:
```diff
@@ -360,6 +385,10 @@ fn read_namespaced_map(chars: &mut std::iter::Enumerate<std::str::Chars>) -> Res
             Some((_, '}')) => return Ok(Edn::NamespacedMap(namespace, Map::new(res))),
+            Some((_, ';')) => {
+                read_comment(chars)?;
+                ()
```
You don't need () here (Rust will interpret a line ending in ; as ().

naomijub commented on src/deserialize/parse.rs:
```diff
@@ -421,6 +454,17 @@ fn read_map(chars: &mut std::iter::Enumerate<std::str::Chars>) -> Result<Edn, Er
 }
 
+fn read_comment(chars: &mut std::iter::Enumerate<std::str::Chars>) -> Result<Edn, Error> {
+    loop {
```
we usually use `chars.take_while(|c| c != \n)`, instead of loops, as it is lazy. Also, have you thought about a test case when the comment doesnt end in a `\n`? I think there is a specific value that you can pass that encompasses new lines, eof, and returns.

jollm replied:
it handles EOF by matching None

I've considered files that use \r instead of \n, probably good to handle that as the spec doesn't emphatically say only \n line endings..

will also change to lazy version

jollm replied:
probably "and all subsequent characters to the next newline should be ignored" is meant to imply newline endings, otherwise it would need to be known whether to expect \r or \n or \r\n?

jollm replied:
Changed to lazy iterator expression.

naomijub commented on src/deserialize/parse.rs:
```diff
@@ -540,6 +604,25 @@ mod test {
 
+    #[test]
+    fn parse_comment_in_simple_vec() {
+        let mut edn = "[11 \"2\" 3.3 ; float in simple vec\n:b true \\c]"
```
I would have some tests that dont actually use `\n` and use strings like:
```
let mut edn = "[
  11 
  \"2\" 
  3.3 
 ;; float in simple vec
 :b 
true 
\\c
]"
```

```

jollm replied:
If I copy that in verbatim it passes. Is that what you intended?

naomijub replied:
Yes

jollm replied:
Added a test like this

naomijub reviewed (changes requested):
Nice works
Thank you for the PR.

jollm committed:
Addressing PR comments

- Return Edn::Empty for empty, whitespace only, and comment only
  strings.
- Fix ; is implied ()
- Use compact lazy iterator expression to read comments
- Add tests for empty and whitespace only cases
- Add a test for literal whitespace ending comment instead of \n
- Add test for string ending in multiple new lines

jollm committed:
Back to top level from parse_edn on comment or whitespace

jollm committed:
Remove unnecessary check for comment and whitespace in parse_edn

Update applicable comment test to call top level parse.

jollm committed:
Smaller change for parsing comments in deserialize

jollm committed:
Remove unnecessary call to map

jollm committed:
Implement commas are whitespace

"Commas , are also considered whitespace, other than within strings."
from https://github.com/edn-format/edn#general-considerations

This simplifies inner parsing levels, no longer having to check for
commas.
